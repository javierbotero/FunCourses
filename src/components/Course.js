import React from 'react';
import { Link } from 'react-router-dom';
import ReactRouterPropTypes from 'react-router-prop-types';
import PropTypes from 'prop-types';

const Course = props => {
  const {
    match,
    courses,
    location,
    url,
  } = props;
  const course = courses.find(c => c.id === parseInt(match.params.id, 10));

  return (
    <div>
      <nav>
        <ul>
          <li>
            <Link to={location.state ? location.state.from.pathname : `${url}`}>
              &#60;
            </Link>
          </li>
        </ul>
      </nav>
      <div>
        <div>{course.title}</div>
        <div>
          <Link
            to={{
              pathname: `${url}/user/${course.teacher_id}`,
              state: { from: location },
            }}
          >
            Teacher:
            {` ${course.teacher.username}`}
          </Link>
        </div>
      </div>
      <div>{course.content}</div>
      <div>
        status:
        {` ${course.status}`}
      </div>
    </div>
  );
};

Course.propTypes = {
  match: ReactRouterPropTypes.match.isRequired,
  location: ReactRouterPropTypes.location.isRequired,
  courses: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    content: PropTypes.string.isRequired,
    teacher_id: PropTypes.number.isRequired,
    status: PropTypes.string.isRequired,
    dates: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    created_at: PropTypes.string.isRequired,
    updated_at: PropTypes.string.isRequired,
    teacher: PropTypes.shape({
      id: PropTypes.number,
      username: PropTypes.string,
    }).isRequired,
    favorites: PropTypes.arrayOf(PropTypes.shape({
      course_id: PropTypes.number,
      user_id: PropTypes.number,
    })).isRequired,
    subscriptions: PropTypes.arrayOf(PropTypes.shape({
      course_id: PropTypes.number,
      user_id: PropTypes.number.isRequired,
      confirmed: PropTypes.bool,
    })).isRequired,
    comments: PropTypes.arrayOf(PropTypes.shape({
      user_id: PropTypes.number,
      course_id: PropTypes.number,
      body: PropTypes.string,
    })).isRequired,
    pendings: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number,
      user_id: PropTypes.number,
      course_id: PropTypes.number,
      confirmed: PropTypes.bool,
    })).isRequired,
  })).isRequired,
  url: PropTypes.string.isRequired,
};

export default Course;
