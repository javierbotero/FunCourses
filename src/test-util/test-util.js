import fetchMock from 'fetch-mock';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { URL } from '../constants/constants';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const mockMatch = {
  params: { id: '5' },
  path: 'some-path',
  url: 'url',
};
const mockCourses = [
  {
    id: 5,
    title: 'Karate class',
    content: 'Nice Karate class',
    teacher_id: 7,
    status: 'Open',
    dates: '2021-03-25T00:00:00.000Z 2021-03-29T00:00:00.000Z ',
    price: 10,
    images_url: [],
    created_at: '2021-03-07T17:56:22.487Z',
    updated_at: '2021-03-15T04:21:24.489Z',
    teacher: {
      id: 7,
      username: 'Adele',
    },
    pendings: [],
    pending_students: [],
    favorites: [],
    subscriptions: [],
    confirmed_students: [],
  },
];
const mockHistory = {
  action: 'PUSH',
  block: jest.fn(),
  createHref: jest.fn(),
  go: jest.fn(),
  listen: jest.fn(),
  push: jest.fn(),
  location: {
    hash: '',
    pathname: '',
    search: '',
  },
  replace: jest.fn(),
};
const mockUser = {
  id: 7,
  status: 'Student',
  username: 'Javier 4',
  url_avatar: 'api.com/users/1',
  courses_as_student: [],
  pending_courses_as_student: [],
  comments: [],
  favorites: [],
  pending_to_accept_friendships: [],
  pending_requested_friendships: [],
  friendship_requests: [],
  requests: [],
  pendings: [],
  courses: [
    {
      id: 5,
      title: 'Karate class',
      content: 'Nice Karate class',
      teacher_id: 7,
      status: 'Open',
      dates: '2021-03-25T00:00:00.000Z 2021-03-29T00:00:00.000Z ',
      price: 10,
      created_at: '2021-03-07T17:56:22.487Z',
      updated_at: '2021-03-15T04:21:24.489Z',
      teacher: {
        id: 7,
        username: 'Adele',
      },
      pendings: [],
      pending_students: [],
      favorites: [],
      subscriptions: [],
      confirmed_students: [],
    },
  ],
};
const mockTeacherCourses = [];
const mockLocation = {
  from: { pathname: 'url' },
  hash: 'some-hash',
  pathname: 'url',
  search: 'default',
  state: {
    user: {
      id: 1,
      username: 'Paco',
    },
    from: {
      pathname: 'url',
    },
  },
};
const mockUrl = 'url';
const mockUsersListToDiv = jest.fn();
const mockCommentsToDivs = jest.fn();
const userPassword = '12345';
const func = jest.fn();
const mockUrlApi = 'url';
const mockToken = 'token';
const mockId = 1;

export {
  func,
  mockStore,
  fetchMock,
  mockMatch,
  mockCourses,
  mockTeacherCourses,
  mockLocation,
  mockUrl,
  mockUsersListToDiv,
  mockCommentsToDivs,
  mockUrlApi,
  mockToken,
  mockId,
  mockUser,
  URL,
  mockHistory,
  userPassword,
};
