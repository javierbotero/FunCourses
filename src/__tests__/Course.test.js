import React from 'react';
import {
  render,
  fireEvent,
  screen,
} from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Course from '../components/Course';
import * as utils from '../test-util/test-util';
import '@testing-library/jest-dom/extend-expect';

describe('Tests for Course', () => {
  it('Checks that', () => {
    render(
      <BrowserRouter>
        <Course
          courses={utils.mockCourses}
          teacherCourses={utils.mockTeacherCourses}
          location={utils.mockLocation}
          url={utils.mockUrl}
          usersListToDiv={utils.mockUsersListToDiv}
          commentsToDivs={utils.mockCommentsToDivs}
          useAuth={utils.mockUseAuth}
          isPresentInUserId={utils.mockIsPresentInUserId}
          isPresentInId={utils.mockIsPresentInId}
          handleLike={utils.mockHandleLike}
          handleSubscription={utils.mockHandleSubscription}
          objThunk={utils.mockObjThunk}
          urlApi={utils.mockUrlApi}
          token={utils.mockToken}
          tokenPayload={utils.mockTokenPayload}
          userPayload={utils.mockUserPayload}
          id={utils.mockId}
          setUserErr={utils.mockSetUserErr}
          match={utils.mockMatch}
        />
      </BrowserRouter>,
    );
    expect(screen.getByText('Karate class')).toHaveTextContent('Karate class');
  });
  it('Checks that', () => {
    render(
      <BrowserRouter>
        <Course
          courses={utils.mockCourses}
          teacherCourses={utils.mockTeacherCourses}
          location={utils.mockLocation}
          url={utils.mockUrl}
          usersListToDiv={utils.mockUsersListToDiv}
          commentsToDivs={utils.mockCommentsToDivs}
          useAuth={utils.mockUseAuth}
          isPresentInUserId={utils.mockIsPresentInUserId}
          isPresentInId={utils.mockIsPresentInId}
          handleLike={utils.mockHandleLike}
          handleSubscription={utils.mockHandleSubscription}
          objThunk={utils.mockObjThunk}
          urlApi={utils.mockUrlApi}
          token={utils.mockToken}
          tokenPayload={utils.mockTokenPayload}
          userPayload={utils.mockUserPayload}
          id={utils.mockId}
          setUserErr={utils.mockSetUserErr}
          match={utils.mockMatch}
        />
      </BrowserRouter>,
    );
    fireEvent.click(screen.getByText('Subscribe'));
    expect(utils.mockHandleSubscription.mock.calls.length).toBe(1);
  });
  it('Checks that', () => {
    render(
      <BrowserRouter>
        <Course
          courses={utils.mockCourses}
          teacherCourses={utils.mockTeacherCourses}
          location={utils.mockLocation}
          url={utils.mockUrl}
          usersListToDiv={utils.mockUsersListToDiv}
          commentsToDivs={utils.mockCommentsToDivs}
          useAuth={utils.mockUseAuth}
          isPresentInUserId={utils.mockIsPresentInUserId}
          isPresentInId={utils.mockIsPresentInId}
          handleLike={utils.mockHandleLike}
          handleSubscription={utils.mockHandleSubscription}
          objThunk={utils.mockObjThunk}
          urlApi={utils.mockUrlApi}
          token={utils.mockToken}
          tokenPayload={utils.mockTokenPayload}
          userPayload={utils.mockUserPayload}
          id={utils.mockId}
          setUserErr={utils.mockSetUserErr}
          match={utils.mockMatch}
        />
      </BrowserRouter>,
    );
    expect(utils.mockUsersListToDiv.mock.calls.length).toBe(2);
  });
  it('Checks that', () => {
    render(
      <BrowserRouter>
        <Course
          courses={utils.mockCourses}
          teacherCourses={utils.mockTeacherCourses}
          location={utils.mockLocation}
          url={utils.mockUrl}
          usersListToDiv={utils.mockUsersListToDiv}
          commentsToDivs={utils.mockCommentsToDivs}
          useAuth={utils.mockUseAuth}
          isPresentInUserId={utils.mockIsPresentInUserId}
          isPresentInId={utils.mockIsPresentInId}
          handleLike={utils.mockHandleLike}
          handleSubscription={utils.mockHandleSubscription}
          objThunk={utils.mockObjThunk}
          urlApi={utils.mockUrlApi}
          token={utils.mockToken}
          tokenPayload={utils.mockTokenPayload}
          userPayload={utils.mockUserPayload}
          id={utils.mockId}
          setUserErr={utils.mockSetUserErr}
          match={utils.mockMatch}
        />
      </BrowserRouter>,
    );
    expect(screen.getByText('Teacher: Adele')).toBeTruthy();
  });
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Course
            courses={utils.mockCourses}
            teacherCourses={utils.mockTeacherCourses}
            location={utils.mockLocation}
            url={utils.mockUrl}
            usersListToDiv={utils.mockUsersListToDiv}
            commentsToDivs={utils.mockCommentsToDivs}
            useAuth={utils.mockUseAuth}
            isPresentInUserId={utils.mockIsPresentInUserId}
            isPresentInId={utils.mockIsPresentInId}
            handleLike={utils.mockHandleLike}
            handleSubscription={utils.mockHandleSubscription}
            objThunk={utils.mockObjThunk}
            urlApi={utils.mockUrlApi}
            token={utils.mockToken}
            tokenPayload={utils.mockTokenPayload}
            userPayload={utils.mockUserPayload}
            id={utils.mockId}
            setUserErr={utils.mockSetUserErr}
            match={utils.mockMatch}
          />
        </BrowserRouter>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
