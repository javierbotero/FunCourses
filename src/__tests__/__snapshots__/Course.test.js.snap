// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Course renders correctly 1`] = `
<div>
  <nav>
    <ul>
      <li>
        <a
          href="/url"
          onClick={[Function]}
        >
          &lt;
        </a>
      </li>
    </ul>
  </nav>
  <div>
    <div>
      Karate class
    </div>
    <div>
      <a
        href="url/user/7"
        onClick={[Function]}
      >
        Teacher:
         Adele
      </a>
    </div>
  </div>
  <div>
    Start:
     3/24/2021
    <br />
    End:
     3/28/2021
  </div>
  <div>
    Nice Karate class
  </div>
  <div>
    status:
     Open
  </div>
  <div>
    price:
     $10
  </div>
  <div>
    <button
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa far,heart"
      />
    </button>
     0
  </div>
  <div>
    <button
      onClick={[Function]}
      type="button"
    >
      
      
      Subscribe
    </button>
  </div>
  <div>
    <div>
      <h4>
        Students confirmed
      </h4>
      <div />
    </div>
    <div>
      <h4>
        Students waiting for confirmation
      </h4>
      <div />
    </div>
  </div>
  <div>
    <h4>
      Comments
    </h4>
    <div />
  </div>
</div>
`;
